javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java

```diff
@@ -1321,6 +1364,127 @@ private static TypeVariable doesNotContainOthers(
+    Context ctx = ((JavacProcessingEnvironment) env).getContext();
+    com.sun.tools.javac.code.Types javacTypes = com.sun.tools.javac.code.Types.instance(ctx);
+    return (ExecutableType) javacTypes.findDescriptorType((Type) functionalInterfaceType);
+  }
+
+  /**
+   * Return whether or not {@code type} is raw.
+   *
+   * @param type the type to check
+   * @return whether or not {@code type} is raw
+   */
+  public static boolean isRaw(TypeMirror type) {
+    if (type.getKind() != TypeKind.DECLARED) {
+      return false;
+    }
+    TypeElement typeelem = (TypeElement) ((DeclaredType) type).asElement();
+    DeclaredType declType = (DeclaredType) typeelem.asType();
+    return !declType.getTypeArguments().isEmpty()
+        && ((DeclaredType) type).getTypeArguments().isEmpty();
+  }
+
+  /**
+   * Returns the most specific supertype of {@code type} that is an array, or null if {@code type}
+   * is not a subtype of an array.
+   *
+   * @param type a type
+   * @param types TypesUtils
+   * @return the most specific supertype of {@code type} that is an array, or null if {@code type}
+   *     is not a subtype of an array
+   */
+  public static @Nullable TypeMirror getMostSpecificArrayType(TypeMirror type, Types types) {
+    if (type.getKind() == TypeKind.ARRAY) {
+      return type;
+    } else {
+      for (TypeMirror superType : types.directSupertypes(type)) {
+        TypeMirror arrayType = getMostSpecificArrayType(superType, types);
+        if (arrayType != null) {
+          return arrayType;
+        }
+      }
+      return null;
+    }
+  }
+
+  /**
+   * Returns true if {@code type} is a parameterized type.
+   *
+   * @param type type to check
+   * @return true if {@code type} is a parameterized type
+   */
+  public static boolean isParameterizedType(TypeMirror type) {
+    return ((Type) type).isParameterized();
+  }
+
+  /**
+   * Return true if {@code typeMirror} is a declared type with at least one wildcard as a type
+   * argument.
+   *
+   * @param typeMirror type to check
+   * @return true if {@code typeMirror} is a declared type with at least one wildcard as a type
+   *     argument
+   */
+  public static boolean isWildcardParameterized(TypeMirror typeMirror) {
+    if (isParameterizedType(typeMirror) && typeMirror.getKind() == TypeKind.DECLARED) {
+      for (TypeMirror t : ((DeclaredType) typeMirror).getTypeArguments()) {
+        if (t.getKind() == TypeKind.WILDCARD) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Creates a wildcard with the given bounds. If {@code upperBound} is Object, then the created
+   * wildcard will not have an upper bound. If {@code upperBound} is {@code null}, then {@code
+   * lowerBound} must not be null.
+   *
+   * @param lowerBound the lower bound for the wildcard
+   * @param upperBound the upper bound for the wilcard
+   * @param types TypesUtils
+   * @return a wildcard with the given bound
```

Comment:
There is no unambiguous "given bound".  This method returns a type with a single bound, but the text above says "Creates a wildcard with the given bound**s**."